woocommerce/woocommerce-ios
# Handle local notifications for trial plan expiration

Part of #9665 


## Description

This PR handles the scheduling of local notifications for trial plan expiration, and the action when the notifications are tapped. Sorry for the many changes, the PR also handles updating push notification manager to make it testable.

Changes include:
- Removed custom actions from `LocalNotification` to use default actions for the local notifications as discussed in pe5sF9-1uQ-p2.
- Updated scenarios `oneDayBeforeFreeTrialExpires` and `oneDayAfterFreeTrialExpires`  to have site ID and updated their identifier accordingly. The reason is that we want to schedule and cancel notifications specific to each site.
- Updated `PushNotesManager` with a new method `requestLocalNotificationIfNeeded`. This method makes sure that only notifications that are not already scheduled can be added to the notification center.
- Updated `UserNotificationsCenterAdapter` and `PushNotificationsManager` for better testability`.
- Updated `StorePlanSynchronizer` to schedule or cancel local notifications based on the fetched site plans.
- Updated `AppCoordinator` to handle the tap on the local notifications.
- Restored `UpgradePlanCoordinatingController` for showing plan upgrade on web view.
- Updated unit tests.

Analytics will be updated in a separate PR.

## Testing instructions

Prerequisites: since the remote feature flags are still not deployed and remain `false`, some code changes are required to test this PR:
- In `LocalNotificationScheduler.schedule`, remove the whole condition on `isRemoteFeatureFlagEnabled(remoteFeatureFlag)` so that the remote feature flag isn't used
- Please also ensure you have access to a site running a trial plan that is not expired yet.

---

- Launch the app, log in, and switch to your test site if needed. For easier testing, please run the app on a physical device.
- Kill the app, open Settings, and change the time of the device to 1 day and 1 minute prior to the expiration date. The expiry date is at 0:00, so my expiry date is on May 19, then I should change the date to May 17, 23:59.
- Wait for 1 minute, then you should see a notification about the expiring trial plan. Tap on the notification, the app should be open, and a web view should be presented to upgrade your plan.
- Kill the app, change the time of your device to 1 day and 59 seconds after the expiration date. E.g: May 19, 23:59, given the above example.
- Wait for 1 minute, then you should see a notification about the expired trial plan. Tap on the notification, the app should be open, and a web view should be presented to upgrade your plan.


## Screenshots

| Expiring plan | Expired plan |
| ----- | ----- |
| <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/9ea0ce0b-91d9-4865-a443-a419efe7ed99" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/d580ec0b-3a1a-4952-a225-eeab9d5c9334" width=320 /> | 
| ----- | ----- |
| <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/0296ffe8-e58b-40e0-af53-e558c38abc50" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/d605b189-27d3-47ea-9f23-1f2929f8dd2d" width=320 /> |





---
